### DISCUSSION TITLE
[Utils] Build a response with empty body

### DISCUSSION THREAD

**Discussion by ar4development** [abhinavsingh/proxy.py#495] [Ideas]
Hi,

I'm using proxy to overtake the chunks and then add header to them. So the method is something like

```
    def handle_upstream_chunk(self, chunk: memoryview) -> memoryview:
        self.response.parse(chunk.tobytes())
        self.response.add_header(b'test_hdr', b'test_val')
        return memoryview(self.response.build_response())
```
However when I get into `self.response.build_response()` with the response having empty body I get `AssertionError`. If to look at that method:

```
    def build_response(self) -> bytes:
        """Rebuild the response object."""
        assert self.code and self.version and self.body and self.type == httpParserTypes.RESPONSE_PARSER
        return build_http_response(
            status_code=int(self.code),
            protocol_version=self.version,
            reason=self.reason,
            headers={} if not self.headers else {
                self.headers[k][0]: self.headers[k][1] for k in self.headers},
            body=self.body if not self.is_chunked_encoded() else ChunkParser.to_chunks(self.body))
```

It asserts that body exists. How do I then build responses having no bodies?

**Comment by abhinavsingh** [comment]
@ar4development Yes, current functions assumes a body will exist.  You can easily modify existing functions or use methods outside of `proxy.py` to build request/response packets.

**Reply by abhinavsingh** [reply]
Opened https://github.com/abhinavsingh/proxy.py/issues/521 to add generic request/response builders in future release.

**Reply by abhinavsingh** [reply]
For posterity, this is now fixed since `v2.4.0`
